rts/lra/lra-test/lra-test-arquillian-extension/src/main/java/io/narayana/lra/arquillian/spi/NarayanaLraRecovery.java

```diff
@@ -0,0 +1,57 @@
+package io.narayana.lra.arquillian.spi;
+
+import io.narayana.lra.LRAConstants;
+import io.narayana.lra.client.NarayanaLRAClient;
+import org.eclipse.microprofile.lra.tck.service.spi.LraRecoveryService;
+
+import javax.enterprise.context.ApplicationScoped;
+import javax.ws.rs.client.Client;
+import javax.ws.rs.client.ClientBuilder;
+import javax.ws.rs.client.WebTarget;
+import javax.ws.rs.core.Response;
+import java.net.URI;
+
+@ApplicationScoped
+public class NarayanaLraRecovery implements LraRecoveryService {
+
+    @Override
+    public void triggerRecovery() {
+        doTriggerRecovery(System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_HOST_KEY),
+            Integer.parseInt(System.getProperty(NarayanaLRAClient.LRA_COORDINATOR_PORT_KEY)));
+    }
+
+    @Override
+    public void triggerRecovery(URI lraId) {
+        doTriggerRecovery(lraId.getHost(), lraId.getPort());
+    }
+
+    private void doTriggerRecovery(String host, int port) {
+        if (!invokeCoordinatorRecovery(host, port)) {
+            // first recovery scan probably collided with periodic recovevery which started
+            // before the test execution so try once more
+            invokeCoordinatorRecovery(host, port);
+        }
+    }
+
+    private boolean invokeCoordinatorRecovery(String host, int port) {
+        // trigger a recovery scan
+        Client recoveryCoordinatorClient = ClientBuilder.newClient();
+
+        try {
+            String recoveryCoordinatorUrl = String.format("http://%s:%d/%s/recovery",
+                host, port, LRAConstants.RECOVERY_COORDINATOR_PATH_NAME);
+            WebTarget recoveryTarget = recoveryCoordinatorClient.target(URI.create(recoveryCoordinatorUrl));
+
+            // send the request to the recovery coordinator
+            Response response = recoveryTarget.request().get();
+            String json = response.readEntity(String.class);
+            response.close();
+
+            // empty JSON array is returned if recovery coordinator doesn't know about any recovering LRA
+            return json.equals("[]");
```

Comment:
This might be just my point of view, but this code is narayana specific so if any previous test fails with active LRAs that is an error on its own and we should add a test to the TCK to verify that is not the case or prune the coordinator object store before every test.

And this line only makes the /recovery endpoint to be called again. Would that be a problem? It will only replay end phase 2 on the pending LRAs from the previous tests -- unnecessary but it shouldn't fail the test.

Reply:
@mmusgrov also can you please respond on google hangouts to @tomjenkinson proposal? It would eliminate the need to call /recovery endpoint from the participant side twice.

Reply:
You are over complicating the fix. Just check whether or not the return JSON contains the LRA id you are looking for, ie just do this:

```java
return json.contains(lraid);
```

Reply:
That looks like it could help to me.

If I followed correctly though, the following test would also need to be made more robust to ensure that it can tolerate forget to be called multiple times? https://github.com/eclipse/microprofile-lra/blob/master/tck/src/main/java/org/eclipse/microprofile/lra/tck/TckContextTests.java#L173

Reply:
no, forget is not an issue here but it wouldn't hurt of course.